### PR TITLE
Speed up frozen actor updates & prevent flicker

### LINKED ISSUE #7660: Buildings frozen under fog flicker
When an enemy building is discovered and after that its visibility is lost, the building's image flickers as it transitions from visible to frozen under fog.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9527505-buildings-frozen-under-fog-flicker?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by RoosterDragon**
Fixes  #7660.

Previously frozen actor updates were done twice per actor. The FrozenUnderFog tick would do this as well as the FrozenActorLayer doing it for every FrozenActor to see if the frozen actor was still visible under the current shroud conditions.

Calculating whether the actor was visible under the shroud is expensive, so we avoid doing it twice since there's no need. Now FrozenUnderFog will update everything as needed.

A couple of further commits introduce some related refactorings and cleanup.

This PR is purely optimization, the behaviour of frozen actors should remain 100% if you're testing this :)

On the RA shellmap, this reduces total CPU time spent doing the frozen actor updates from 6.21% to 4.13%.

### CODE REVIEW

**Comment by Mailaender** [comment]
This regresses badly:

image

The left pill box and wall segments have been scouted, fogged, then destroyed by artillery. Revealing them again makes them still show up in a glitchy manner.

**Comment by RoosterDragon** [comment]
Yup that's broken alright - what an obvious thing to not occur to me to test!

Anyway this seems to be caused by the fact FrozenUnderFog stops ticking once the actor dies. So it we turn the logic the other way around - FrozenActorLayer does the work and FrozenUnderFog uses the result - that fixes the problem since the layer keeps on ticking.

This makes the diff a bit uglier, but the performance improvement is the same.

**Comment by pchote** [comment]
This may also fix the flicker when switching between the regular and frozen actor rendering, which would be a great unexpected side effect.

**Comment by RoosterDragon** [comment]
I'd never noticed that was an issue before - interesting.

Just tested it, and yes it seems I (totally unintentionally) fixed it - how fortuitous.

File: OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs [hunk 1]
```diff
@@ -66,55 +65,43 @@ public void Tick(Actor self)
...
 			{
-				foreach (var p in self.World.Players)
+				bool isVisible;
+				FrozenActor frozenActor;
+				if (!initialized)
 				{
-					visible[p] |= startsRevealed;
-					p.PlayerActor.Trait<FrozenActorLayer>().Add(frozen[p] = new FrozenActor(self, footprint, footprintRegion));
+					frozen[player] = frozenActor = new FrozenActor(self, footprint, footprintRegion, player.Shroud);
+					player.PlayerActor.Trait<FrozenActorLayer>().Add(frozenActor);
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs]: You could cache the trait lookups for all `self.World.Players` in the constructor to save even more resources.

Comment (RoosterDragon) [inline OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs]: The trait lookup only happens on the first tick so it's not an issue.

Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Modifiers/FrozenUnderFog.cs]: You are right.

**Comment by Mailaender** [comment]
:+1: The regressions are fixed. Tested both revealing, fogging, destroying, revealing and starting a map without shroud and just fog where tech buildings are correctly revealed at start, but others not. :white_check_mark:

**Comment by pchote** [comment]
I couldn't spot anything wrong with the code or ingame, so :+1: from me too.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/2887629d44b3f2083ac030ad26969a228dbaed57%5E...2887629d44b3f2083ac030ad26969a228dbaed57

**Pull request merged**